### PR TITLE
added brunel network tutorial

### PR DESCRIPTION
**Pull request by ErbB4**
added Brunel network tutorial on Arbor

Closes: #1933

### CODE REVIEW

File: doc/tutorial/brunel.rst [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+   :dedent:
+   :lines: 126-178
+
+One can also use the code below to visualize the raster plot of the entire nework and a few selected cells, and the peristimulus time histogram (PSTH) of the entire network. The parameters used here are supposed to achieve asynchronous irregular dynamics.
+
+.. literalinclude:: ../../python/example/brunel/analysis.py
+   :language: python
+   :dedent:
+   :lines: 1-60
+
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:72]: Show the output / plots here, please. You can save them as SVG and steal the include from LIF
```

.. figure:: ../images/lif.svg
    :width: 600
    :align: center

    Plot of the potential over time for a LIF cell.
```

File: doc/tutorial/brunel.rst [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+The full code
+-------------
+You can find the same network architecture simulated in the NEST simulator in the same repo ``python/examples/brunel/nest_brunel.py``. The average firing rate of neurons and network dynamics look similar in both cases.
+
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:77]: Also here, output / plot appreciated.

File: doc/tutorial/brunel.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+The Brunel network is randomly sparsely connected with a fixed in-degree regulated by a connection probability (:math:`\epsilon`). We, therefore, define a function to enable random connectivity.
+
+.. literalinclude:: ../../python/example/brunel/arbor_brunel.py
+   :language: python
+   :dedent:
+   :lines: 93-112
+
+
+To enable the network activity, we apply Poissonian input via event_generatorto to each neuron in the network. It aims to achieve a similar effect as the `Poisson_generator` in the NEST simulator.
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:51]:
```suggestion
To enable the network activity, we apply Poissonian input via `event_generators` to to each neuron in the network. It aims to achieve a similar effect as the `Poisson_generator` in the NEST simulator.
```

File: doc/tutorial/brunel.rst [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+We define the network size with **num_cells** and cell type with **cell_kind**. Then load all the neuron parameters to the LIF cells with **cell_description** function.
+
+.. literalinclude:: ../../python/example/brunel/arbor_brunel.py
+   :language: python
+   :dedent:
+   :lines: 62-81
+
+
+The Brunel network is randomly sparsely connected with a fixed in-degree regulated by a connection probability (:math:`\epsilon`). We, therefore, define a function to enable random connectivity.
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:43]: A short explanation what `sample_subset` aims to achieve?

File: doc/tutorial/brunel.rst [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ 1. Build LIF cells by loading certain neuron parameters from the parameter file.
+ 2. Connect neurons in a fixed in-degree manner based on a connection probability.
+ 3. Add Possonian input to drive the network activity.
+ 4. Record spikes and plot raster plot and peristimulus time histogram (PSTH).
+
+
+Start a recipe and initiate the parameters
+------------------------------------------
+
+Here, we will follow the protocol for building the recipe, which has been instructed before.
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:20]:
```suggestion
Here, we will follow the protocol for building the recipe, which has been discussed before.
```
A link to the single cell recipe example or the ring network would be good, too

File: doc/tutorial/brunel.rst [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+==============
+
+In this tutorial, we will build a classic Brunel network using LIF cells in Arbor, and you can compare it with the simulation using LIF cells in the NEST simulator.
+
+.. Note::
+    
+    **Concepts covered in this example:**
+ 1. Build LIF cells by loading certain neuron parameters from the parameter file.
+ 2. Connect neurons in a fixed in-degree manner based on a connection probability.
+ 3. Add Possonian input to drive the network activity.
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:13]:
```suggestion
 3. Add Poissonian input to drive the network activity.
```

File: python/example/brunel/analysis.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+def analysis_rate(sources,times,tfinal):
+    rates = []
+
+    for neuron in np.arange(0,NE+NI,1):
+        idx = np.where(sources==neuron+1)
+        time = times[idx]
+        rate = len(time)/(tfinal/1000.)
+        rates.append(rate)
+    print(np.mean(rates[0:NE]))
```
Comment (thorstenhater) [inline python/example/brunel/analysis.py:38]: Needed?

File: python/example/brunel/analysis.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    plt.hist(rates)
+    plt.xlabel("rate | Hz")
+    plt.ylabel("counts")
+
+
+times = np.load("times.npy")
+sources = np.load("sources.npy")
+plot_raster_and_PSTH(times,sources)
+plot_rate_histogram(sources,times,tfinal)
+plt.show()
```
Comment (thorstenhater) [inline python/example/brunel/analysis.py:56]: Probably better save as a PDF and use later? At least that's how all the other tutorials are made.

File: python/example/brunel/nest_brunel.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+import parameters
+from parameters import *
+import nest
+
+
+
+
+#set up seed for random number
+seed        = 0*6748+7928
```
Comment (thorstenhater) [inline python/example/brunel/nest_brunel.py]: Seed not taken from parameters?

**Review by thorstenhater — CHANGES_REQUESTED** [review]
Finally someone does this. :D

File: python/example/brunel/analysis.py (hunk 7 above)
Comment (ErbB4) [reply python/example/brunel/analysis.py:38]: yes

File: python/example/brunel/analysis.py (hunk 8 above)
Comment (ErbB4) [reply python/example/brunel/analysis.py:56]: saved as svg

File: python/example/brunel/analysis.py (hunk 7 above)
Comment (thorstenhater) [reply python/example/brunel/analysis.py:44]: It'll the result to the terminal without telling the user what it is.
So, I'd suggest one of these
1. Drop it
2. Add a statistics block at the end, printing various summary statistics.
3. Add it to the plot as a text label.

File: doc/tutorial/brunel.rst [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+We define the network size with **num_cells** and cell type with **cell_kind**. Then load all the neuron parameters to the LIF cells with **cell_description** function.
+
+.. literalinclude:: ../../python/example/brunel/arbor_brunel.py
+   :language: python
+   :dedent:
+   :lines: 62-81
+
+
+The Brunel network is randomly sparsely connected with a fixed in-degree regulated by a connection probability (:math:`\epsilon`). We, therefore, define a function to enable random connectivity. This funciton draws random connections from the pre-selected pool of source neruons defined by the `gid` defined within `start` and `end`. The total number of random connections are regulated by the fix-indegree value, which is `m` here in the function and `CE` and `CI` in the pamaters file.
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:44]:
```suggestion
The Brunel network is randomly sparsely connected with a fixed in-degree regulated by a connection probability (:math:`\epsilon`). We, therefore, define a function to enable random connectivity. This funciton draws random connections from the pre-selected pool of source neruons defined by the `gid` defined within `start` and `end`. The total number of random connections are regulated by the fix-indegree value, which is `m` here in the function and `CE` and `CI` in the parameters file.
```

File: doc/tutorial/brunel.rst [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+Here, we will follow the protocol for building the recipe, which has been discussed before in :ref:`Ring Network
+<tutorialnetworkring>`.
+
+.. literalinclude:: ../../python/example/brunel/arbor_brunel.py
+   :language: python
+   :dedent:
+   :lines: 4-59
+
+We can find all the parameters listed in a seperate parameter files:
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:28]:
```suggestion
We can find all the parameters listed in a separate parameter file:
```

File: doc/tutorial/brunel.rst [hunk 12]
```diff
@@ -0,0 +1,113 @@
+.. _tutorialbrunel:
+
+Brunel network
+==============
+
+In this tutorial, we will build a classic Brunel network using LIF cells in Arbor, and you can compare it with the simulation using LIF cells in the NEST simulator.
+
+.. Note::
```
Comment (thorstenhater) [inline doc/tutorial/brunel.rst:8]:
```suggestion
.. admonition:: Concepts covered in this example
   :class: note
```

and drop the line below

**Review by thorstenhater — CHANGES_REQUESTED** [review]
Almost there, a few comments:

1. The images do not render in the preview build, something is off there. See https://arbor--2387.org.readthedocs.build/en/2387/tutorial/brunel.html#the-full-code
2. Lint is failing due to the `####` marks. I'd suggest removing those.
3. Lint also doesn't like `import *`, but here it's ok. I'll find a way to disable it, if you like. Alternatively, you could use a really short name, e.g. `import parameters as P`.

File: file.text
Comment (thorstenhater) [inline file.text:1]: This one shouldn't be here, right?

File: .flake8 [hunk 13]
```diff
@@ -23,3 +23,7 @@ extend-exclude =
                build,
                # nah, don't care
                spack/package.py
+per-file-ignores=
+    python/example/brunel/analysis.py=F405,F403
```
Comment (thorstenhater) [inline .flake8:27]:
```suggestion
    python/example/brunel/analysis.py:F405,F403
```
and the next two lines.

**Review by thorstenhater — CHANGES_REQUESTED** [review]

File: python/example/brunel/parameters.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
+import numpy as np
+
+
+#### define network parameters ####
```
Comment (thorstenhater) [inline python/example/brunel/parameters.py:4]:
```suggestion
# define network parameters
```

**Review by thorstenhater — APPROVED** [review]

**Pull request merged**